drivers/meter/ade7913/ade7913.c

```diff
@@ -0,0 +1,2308 @@
+	};
+
+	if (init_param.drdy_callback)
+		irq_cb.callback = init_param.drdy_callback;
+
+	/* SPI Initialization */
+	ret = no_os_spi_init(&dev->spi_desc, init_param.spi_init);
+	if (ret)
+		goto error_dev;
+
+	ret = no_os_gpio_get_optional(&dev->gpio_rdy, init_param.gpio_rdy);
+	if (ret)
+		goto error_spi;
+
+	if (dev->gpio_rdy) {
+		ret = no_os_gpio_direction_input(dev->gpio_rdy);
+		if (ret)
+			goto error_gpio;
+	}
+
+	ret = no_os_irq_register_callback(init_param.irq_ctrl,
+					  dev->gpio_rdy->number, &irq_cb);
+	if (ret)
+		goto error_gpio;
+
+	dev->irq_cb = irq_cb;
+
+	ret = no_os_irq_trigger_level_set(init_param.irq_ctrl,
+					  dev->gpio_rdy->number, NO_OS_IRQ_EDGE_FALLING);
+	if (ret)
+		goto error_irq;
+
+	ret = no_os_irq_set_priority(init_param.irq_ctrl, dev->gpio_rdy->number, 3);
+	if (ret)
+		goto error_irq;
+
+	ret = no_os_irq_disable(init_param.irq_ctrl,
+				dev->gpio_rdy->number);
+	if (ret)
+		goto error_irq;
+
+	dev->irq_ctrl = init_param.irq_ctrl;
+
+	ret = no_os_gpio_get_optional(&dev->gpio_reset, init_param.gpio_reset);
+	if (ret)
+		goto error_spi;
+
+	if (dev->gpio_reset) {
+		ret = no_os_gpio_direction_output(dev->gpio_reset, NO_OS_GPIO_HIGH);
+		if (ret)
+			goto error_gpio;
+	}
+
+	/* Create the CRC-8 lookup table for polynomial ADE7913_CRC8_POLY */
+	no_os_crc8_populate_msb(ade7913_crc8, ADE7913_CRC8_POLY);
+
+	/* Create the CRC-16 lookup table for polynomial ADE7913_CRC8_POLY */
+	no_os_crc16_populate_msb(ade7913_crc16, ADE7913_CRC16_POLY);
+
+	/* CRC enabled by default */
+	dev->crc_en = 1;
+
+	/* Reset device */
+	ret = no_os_gpio_set_value(dev->gpio_reset, NO_OS_GPIO_LOW);
+	if (ret)
+		goto error_gpio;
+
+	no_os_mdelay(10);
+
+	ret = no_os_gpio_set_value(dev->gpio_reset, NO_OS_GPIO_HIGH);
+	if (ret)
+		goto error_gpio;
+
+	no_os_mdelay(50);
+
+	do {
+		ret = ade7913_get_com_up(dev, &reg_val);
+		if (ret)
+			goto error_gpio;
+
+		no_os_mdelay(5);
+		timeout++;
+		if (timeout == 20) {
+			ret = -ENOTCONN;
+			goto error_gpio;
+		}
+	} while (!reg_val);
+
+	/* Read version product */
+	ret = ade7913_get_version_product(dev, &reg_val);
+	if (ret)
+		goto error_gpio;
+
+	dev->ver_product = reg_val;
+
+	/* Read silicon revision */
+	ret = ade7913_get_silicon_revision(dev, &reg_val);
+	if (ret)
+		goto error_gpio;
+
+	if (reg_val != ADE7913_SILICON_REVISION) {
+		ret = -ENODEV;
+		goto error_gpio;
+	}
+
+	*device = dev;
+
+	return 0;
+
+error_irq:
+	no_os_irq_unregister_callback(init_param.irq_ctrl, dev->gpio_rdy->number,
+				      &irq_cb);
+error_gpio:
+	no_os_gpio_remove(dev->gpio_rdy);
+error_spi:
+	no_os_spi_remove(dev->spi_desc);
+error_dev:
+	no_os_free(dev);
+
+	return ret;
+}
+
+/**
+ * @brief Remove the device and release resources.
+ * @param dev - The device structure.
+ * @return 0 in case of success, negative error code otherwise.
+ */
+int ade7913_remove(struct ade7913_dev *dev)
+{
+	int ret;
+
+	ret = no_os_spi_remove(dev->spi_desc);
+	if (ret)
+		return ret;
+
+	no_os_free(dev);
```

Comment:
I think there are some memory leaks here. no_os_gpio_get_optional and no_os_irq_register_callback allocate memory which should be freed

Reply:
Fixed.